### PR TITLE
Validator allowlist

### PR DESCRIPTION
**Pull request by akorchyn**
Features:
- [x] restrict only chosen calls
- [x] make validators able to vote to add a new entity to whitelist
- [x] remove ability to withdraw entity from whitelist
- [x] cover with tests

### CODE REVIEW

**Comment by akorchyn** [comment]
@xcthulhu let's protect `main` branch

Update: done

**Review by uchu — DISMISSED** [review]
lgtm

File: frame/account-filter/src/lib.rs [hunk 1]
```diff
@@ -87,40 +105,56 @@ pub mod pallet {
...
-                Error::<T>::AccountNotAdded
+                !<Votes<T>>::contains_key(&new_account, &account),
+                Error::<T>::DuplicateVote
             );
 
-            <AllowedAccounts<T>>::remove(&account_to_remove);
-
-            Self::deposit_event(Event::AccountRemoved(account_to_remove));
+            // Check if the new account has enough votes to be added to the allow-list.
+            let votes_for = Votes::<T>::iter_prefix(&new_account).count() + 1;
```
Comment (obsessed-cake) [inline frame/account-filter/src/lib.rs]:
```suggestion
            let votes_for = Votes::<T>::iter_prefix(&new_account)
                .count()
                .checked_add(1)
                .ok_or(Error::<T>::VotesCounterOverflow)?;
```

File: frame/account-filter/src/lib.rs [hunk 2]
```diff
@@ -185,25 +218,25 @@ pub mod pallet {
...
             who: &Self::AccountId,
-            _call: &Self::Call,
+            call: &Self::Call,
             info: &DispatchInfoOf<Self::Call>,
             _len: usize,
         ) -> TransactionValidity {
-            if <pallet::AllowedAccounts<T>>::contains_key(who) {
+            if T::CallsToFilter::matches(call) && !<AllowedAccounts<T>>::contains_key(who) {
+                Err(InvalidTransaction::BadSigner.into())
+            } else {
```
Comment (obsessed-cake) [inline frame/account-filter/src/lib.rs]:
```suggestion
            ensure!(
                !T::CallsToFilter::matches(call) && <AllowedAccounts<T>>::contains_key(who),
                InvalidTransaction::BadSigner
            );
```

Comment (akorchyn) [reply frame/account-filter/src/lib.rs]: You probably meant: !T::CallsToFilter::matches(call) **||** <AllowedAccounts<T>>::contains_key(who)

File: frame/account-filter/src/lib.rs [hunk 3]
```diff
@@ -31,31 +36,44 @@ pub mod pallet {
...
     #[pallet::getter(fn allowed_accounts)]
     pub type AllowedAccounts<T: Config> = StorageMap<_, Blake2_128Concat, T::AccountId, ()>;
 
+    // Voting process for the allow-list.
+    // The key is the account that is being voted for. The value is the account that is voting for.
+    #[pallet::storage]
+    #[pallet::getter(fn votes)]
+    pub type Votes<T: Config> =
+        StorageDoubleMap<_, Blake2_128Concat, T::AccountId, Blake2_128Concat, T::AccountId, ()>;
+
```
Comment (obsessed-cake) [inline frame/account-filter/src/lib.rs:61]: Also we can add
```
#[pallet::storage]
#[pallet::getter(fn votes_for_account)]
pub type VotesForAccount<T: Config> = StorageMap<_, Blake2_128Concat, T::AccountId, u128>;
```
to count votes faster.

File: frame/account-filter/src/lib.rs (hunk 2 above)
Comment (obsessed-cake) [reply frame/account-filter/src/lib.rs]: Yes, so.

File: frame/account-filter/src/lib.rs (hunk 1 above)
Comment (akorchyn) [reply frame/account-filter/src/lib.rs]: done

File: frame/account-filter/src/lib.rs (hunk 2 above)
Comment (akorchyn) [reply frame/account-filter/src/lib.rs]: done

File: frame/account-filter/src/lib.rs (hunk 3 above)
Comment (akorchyn) [reply frame/account-filter/src/lib.rs:61]: Done. Also, made for allowlisted users.

**Review by baidang201 — DISMISSED** [review]

**Review by obsessed-cake — APPROVED** [review]

**Comment by obsessed-cake** [comment]
LGTM

**Pull request merged**